node/src/components/linear_chain_sync/operations.rs

```diff
@@ -473,17 +478,26 @@ async fn fetch_and_store_block_by_hash(
     }
 }
 
-/// Synchronize the trie store under a given state root hash.
+/// Synchronizes the trie store under a given state root hash.
 async fn sync_trie_store(
     effect_builder: EffectBuilder<JoinerEvent>,
     state_root_hash: Digest,
+    max_parallel_trie_fetches: usize,
 ) -> Result<(), LinearChainSyncError> {
     info!(?state_root_hash, "syncing trie store",);
+    // TODO: This implementation works like a stream with ordered buffering. Use an actual stream
+    //       with unordered buffering instead.
     let mut outstanding_trie_keys = vec![state_root_hash];
-    while let Some(trie_key) = outstanding_trie_keys.pop() {
-        let missing_descendant_trie_keys =
-            fetch_trie_and_insert_into_trie_store(effect_builder, trie_key).await?;
-        outstanding_trie_keys.extend(missing_descendant_trie_keys);
+    let mut fetches_in_progress = VecDeque::with_capacity(max_parallel_trie_fetches);
+    while !fetches_in_progress.is_empty() || !outstanding_trie_keys.is_empty() {
+        let fetches_to_start = max_parallel_trie_fetches.saturating_sub(fetches_in_progress.len());
+        let new_fetches = outstanding_trie_keys
+            .drain(outstanding_trie_keys.len().saturating_sub(fetches_to_start)..)
```

Comment:
This will always start fetching from the end of the queue, which I think means we'll fetch depth-first. Is that intentional? (Not sure if it matters at all TBH - probably not.)

Reply:
I don't think it matters, but if there's an argument to fetch from the beginning instead I'm happy to change it.